### ISSUE TITLE
Do error reporting in twisted.names.client.Resolver

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2248]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2248|
|Type|task|
|Created|2006-11-28 16:41:33Z|

Resolver only notices truncation errors.  It doesn't check the result code to determine if any other failure occurred.  This makes it difficult or impossible to differentiate between different failure conditions.

<details><summary>Searchable metadata</summary>

```
trac-id__2248 2248
type__task task
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1164732093000000 1164732093000000
changetime__1165374938000000 1165374938000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @jml|
|-|-|

Ready for review in resolver-error-reporting-2248

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @radix<br>@radix set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_3"></a>|@radix set owner to @exarkun<br>@radix set status to `new`|
|-|-|

+        """
+        Test that if the MX record lookup fails, fallback is enabled, and an A
+        record is available for the name, Deferred returned by
+        L{MXCalculator.getMX} ultimately fires with a Record_MX instance which
+        gives the address in the A record for the name.
+        """


I don't understand what's going on in that English after the comma following "name".

80 columns.


    def _cbMX(self, answers, domain):
        if not answers:
            raise IOError("No MX found for %r" % (domain,))

Is that necessary any more?



OK, please respond to these and then probably merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `closed`|
|-|-|

(In [18817]) Merge resolver-error-reporting-2248

Author: exarkun
Reviewer: radix
Fixes [#2248](https://github.com/twisted/twisted/issues/2248)

Change the resolver client API to translate DNS response codes other than OK
into Failures passed to a Deferred errback.  Add several new exception classes
to represent the different response codes which are possible and a new module
for twisted.names exception classes.

This is a not a backwards compatible change.  Instead of being passed a three
tuple of empty lists on error, callers of lookup methods will now have their
Deferred errback instead.

Also update twisted.mail.relaymanager.MXCalculator to account for this change,
since it relied on checks against the empty list to discover the failure case.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
